File: LeetCode/Easy/125. Valid Palindrome.py
# A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters and numbers.

# Given a string s, return true if it is a palindrome, or false otherwise.

 

# Example 1:

# Input: s = "A man, a plan, a canal: Panama"
# Output: true
# Explanation: "amanaplanacanalpanama" is a palindrome.
# Example 2:

# Input: s = "race a car"
# Output: false
# Explanation: "raceacar" is not a palindrome.
# Example 3:

# Input: s = " "
# Output: true
# Explanation: s is an empty string "" after removing non-alphanumeric characters.
# Since an empty string reads the same forward and backward, it is a palindrome.
 

# Constraints:

# 1 <= s.length <= 2 * 105
# s consists only of printable ASCII characters.

# My Python First Solution of O(n) time complexity

class Solution:
    def isPalindrome(self, s: str) -> bool:
        alphabet_dict: dict(str) = {'a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u', 'v','w','x','y','z', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9'}
        i_start: int = 0
        i_end: int = len(s) - 1

        while i_start < i_end:
            print(s[i_start] + ", " + s[i_end])
            print(str(i_start) + ", " + str(i_end))
            
            while not s[i_start].lower() in alphabet_dict:
                i_start += 1

                try:
                    s[i_start]
                except IndexError:
                    return True

            while not s[i_end].lower() in alphabet_dict:
                i_end -= 1

                try:
                    s[i_end]
                except IndexError:
                    return True
                    

            if s[i_start].lower() != s[i_end].lower():
                return False

            i_start += 1
            i_end -= 1


        
        return True
    
# My 2nd Pass Solution

class Solution:
    def isPalindrome(self, s: str) -> bool:
        i_start: int = 0
        i_end: int = len(s) - 1

        while i_start < i_end:            
            while not self.isAlphanumeric(s[i_start].lower()):
                i_start += 1

                try:
                    s[i_start]
                except IndexError:
                    return True

            while not self.isAlphanumeric(s[i_end].lower()):
                i_end -= 1

                try:
                    s[i_end]
                except IndexError:
                    return True
                    

            if s[i_start].lower() != s[i_end].lower():
                return False

            i_start += 1
            i_end -= 1


        
        return True

    def isAlphanumeric (self, ch: str) -> bool:
        return (ord(ch) <= 57 and ord(ch) >= 48 ) or (ord(ch) <= 122 and ord(ch) >= 97)

# My 3rd Pass Solution
class Solution:
    def isPalindrome(self, s: str) -> bool:
        i_start: int = 0
        i_end: int = len(s) - 1

        while i_start < i_end:      

            if not self.isAlphanumeric(s[i_start].lower()):
                i_start += 1
                continue

            elif not self.isAlphanumeric(s[i_end].lower()):
                i_end -= 1
                continue


            elif s[i_start].lower() != s[i_end].lower():
                return False

            else:
                i_start += 1
                i_end -= 1


        
        return True

    def isAlphanumeric (self, ch: str) -> bool:
        return (ord(ch) <= 57 and ord(ch) >= 48 ) or (ord(ch) <= 122 and ord(ch) >= 97)


# My Final Pass Solution:
class Solution:
    def isPalindrome(self, s: str) -> bool:
        i_start: int = 0
        i_end: int = len(s) - 1

        while i_start < i_end:      

            if not s[i_start].isalnum():
                i_start += 1

            elif not s[i_end].isalnum():
                i_end -= 1


            elif s[i_start].lower() != s[i_end].lower():
                return False

            else:
                i_start += 1
                i_end -= 1


        
        return True
    
# Leetcode Comunity Alternative

class Solution:
    def isPalindrome(self, s: str) -> bool:
        processed_s = re.sub(r'[^a-zA-Z0-9]', '', s).lower()
        return processed_s == processed_s[::-1]

File: LeetCode/Easy/125. Valid Palindrome.cpp
// A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters and numbers.

// Given a string s, return true if it is a palindrome, or false otherwise.

 

// Example 1:

// Input: s = "A man, a plan, a canal: Panama"
// Output: true
// Explanation: "amanaplanacanalpanama" is a palindrome.
// Example 2:

// Input: s = "race a car"
// Output: false
// Explanation: "raceacar" is not a palindrome.
// Example 3:

// Input: s = " "
// Output: true
// Explanation: s is an empty string "" after removing non-alphanumeric characters.
// Since an empty string reads the same forward and backward, it is a palindrome.
 

// Constraints:

// 1 <= s.length <= 2 * 105
// s consists only of printable ASCII characters.

// My C++ Solution O(n)

class Solution {
public:
    bool isPalindrome(string s) {
        const char* head = s.c_str();
        const char* tail = &s.back();

        while (tail >= head) {

            while(tail >= head && !isalnum(*tail)) {
                --tail;
            }

             while(tail >= head && !isalnum(*head)) {
                ++head;
            }

            if(!isalnum(*head) || !isalnum(*tail)) return true;
            if(std::tolower(*tail) != std::tolower(*head)) return false;

            ++head;
            --tail;
        }

        return true;
    }
};

// Leetcode Comunity Alternative

class Solution {
public:
    
bool isaLnum(char c){
    return(c >='A' && c<= 'Z')||
           (c >= 'a' && c<= 'z')||
           (c >= '0' && c <= '9');
}
char toLower(char c){
    if(c>= 'A'&& c<='Z'){
        return c + ('a'-'A');
    }
    return c;
}
bool isPalindrome(std::string s){
    int left = 0, right = s.length()-1;
    while(left<right){
        while(left<right && !isaLnum(s[left])) left++;
        while(left<right && !isaLnum(s[right])) right--;

        if(toLower(s[left])!=toLower(s[right])){
            return false;
        }
        left++;
        right--;
    }
    return true;
}
int main(){
    string str;
    cin >> str;
    if(isPalindrome(str)){
        cout << "Is an palindrome";
    }
    else cout << "not an palindrome!";
    return 0;
}
};


// Or

#include <cctype> // for isalnum and tolower

class Solution {
public:
    bool isPalindrome(string s) {
        int i = 0, j = s.size() - 1;

        while (i < j) {
            // Skip non-alphanumeric characters
            while (i < j && !isalnum(s[i])) i++;
            while (i < j && !isalnum(s[j])) j--;

            // Compare lowercase characters
            if (tolower(s[i]) != tolower(s[j])) return false;

            i++;
            j--;
        }

        return true;
    }
};